File: implementation/c#/p20-talos/Authorizer.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using p20_talos.tree;

namespace p20_talos
{
    public class Authorizer
    {
        private readonly Dictionary<string, Node> _permissions = new Dictionary<string, Node>();

        private readonly Dictionary<string, string> _inheritance = new Dictionary<string, string>();

        public bool HasAccess(string role, string resource, Dictionary<string, string> variables,
            Dictionary<string, HashSet<string>> sets)
        {
            if (!_permissions.ContainsKey(role))
            {
                return false;
            }
            
            if (resource[0] == '/')
            {
                resource = resource.Substring(1);
            }

            bool? hasAccess = _permissions[role].HasAccess(resource, variables, sets) ??
                              _inheritance.ContainsKey(role) &&
                              HasAccess(_inheritance[role], resource, variables, sets);

            return hasAccess.Value;
        }

        public void LoadPermissions(string data)
        {
            var doubleSpacesRegex = new Regex("[ ]{2,}", RegexOptions.None);
            data.Split('\n')
                .Select(r => r.Trim())
                .Where(r => r.Length > 0 && r[0] != '#')
                .Select(r => doubleSpacesRegex.Replace(r, " ").Split(' ').Select(p => p.Trim()).ToArray())
                .ToList().ForEach(r =>
                {
                    CheckRuleSyntaxis(r);
                    
                    var ruleType = GetRuleType(r);

                    if (ruleType == RuleType.Inheritance)
                    {
                        SetInheritance(r);
                    }
                    else
                    {
                        SetPermission(r);
                    }
                });
        }

        private static void CheckRuleSyntaxis(string[] rule)
        {
            if (rule.Length != 3)
            {
                throw new ParseException("Malformed rule '" + string.Join(" ", rule) + "': a rule should have three parts.");
            }
        }

        private static RuleType GetRuleType(IReadOnlyList<string> rule)
        {
            var isInheritance = rule[1] == ">";
            var isPermission = (rule[0] == "allow" || rule[0] == "deny");

            if (isInheritance)
            {
                if (isPermission)
                {
                    throw new ParseException(
                        "Ambiguous rule: a rule cannot be both a permission rule and an inheritance rule.");
                }

                return RuleType.Inheritance;
            }

            if (isPermission)
            {
                return RuleType.Permission;
            }

            throw new ParseException(
                "Malformed rule: a rule should either be a permission rule or a inheritance rule.");
        }

        private void SetInheritance(IReadOnlyList<string> rule)
        {
            _inheritance[rule[2]] = rule[0];
        }

        private void SetPermission(IReadOnlyList<string> rule)
        {
            if (!_permissions.ContainsKey(rule[1]))
            {
                _permissions[rule[1]] = new RootNode();
            }

            var path = rule[2].Substring(rule[2][0] == '/' ? 1 : 0);

            _permissions[rule[1]].AddRule(rule[0] == "allow", path);
        }

        public void LoadPermissionsFile(string filename)
        {
            LoadPermissions(File.ReadAllText(filename));
        }

        public void ClearPermissions()
        {
            _inheritance.Clear();
            _permissions.Clear();
        }

        private enum RuleType
        {
            Inheritance, Permission
        }
    }
}

File: implementation/php/src/Authorizer.php
<?php

include "nodes.php";

class Authorizer {

	public $preconstruct;
	private $permissionRules = "";

	private $inheritance = [];
	private $permissions = [];

	function __construct($preconstruct = true) {
		$this->preconstruct = $preconstruct;
	}

	function hasAccess($role, $path, $variables, $sets) {
		if (!isset($this->permissions[$role])) {
			return false;
		}

		if (substr($path, 0, 1) === '/') {
			$path = substr($path, 1);
		}

		$hasAccess = $this->permissions[$role]->hasAccess($path, $variables, $sets);

		if ($hasAccess !== null) {
			return $hasAccess;
		}

		if (isset($this->inheritance[$role])) {
			return $this->hasAccess($this->inheritance[$role], $path, $variables, $sets);
		} else {
			return false;
		}
	}


	function loadPermissionsFile($filename) {
		$this->loadPermissions(file_get_contents($filename));
	}

	function loadPermissions($permissions) {
		$this->permissionRules .= PHP_EOL . $permissions;

		if ($this->preconstruct) {
			$this->parsePermissions();
		}
	}

	private function parsePermissions() {
		foreach (array_map(
								 function ($r) {
									 return array_map(
											 function ($word) {
												 return trim($word);
											 },
											 explode(" ", preg_replace("/[ ]{2,}/", " ", $r))
									 );
								 },
								 array_filter(
										 array_map(
												 function ($r) {
													 return trim($r);
												 },
												 explode("\n", $this->permissionRules)
										 ),
										 function ($r) {
											 return strlen($r) > 0 && substr($r, 0, 1) !== "#";
										 }
								 )
						 ) as $rule) {
			$this->checkRuleSyntax($rule);

			$ruleType = $this->getRuleType($rule);

			if ($ruleType === "Inheritance") {
				$this->setInheritance($rule);
			} else {
				$this->setPermission($rule);
			}
		}

		$this->permissionRules = "";
	}

	private function checkRuleSyntax($rule) {
		if (count($rule) != 3) {
			throw new MalformedRuleException("The rule \"" . join(", ", $rule) . "\" does not have three pieces.");
		}
	}

	private function getRuleType($rule) {
		$isInheritance = $rule[1] == ">";
		$isPermission = ($rule[0] === "allow" || $rule[0] === "deny");

		if ($isInheritance) {
			if ($isPermission) {
				throw new MalformedRuleException("The type of the rule \"" . join(", ", $rule) . "\" is ambiguous.");
			} else {
				return "Inheritance";
			}
		} else {
			if ($isPermission) {
				return "Permission";
			} else {
				throw new MalformedRuleException("The rule \"" . join(", ", $rule) . "\" is neither a inheritance rule nor a permission rule.");
			}
		}
	}

	private function setInheritance($rule) {
		$this->inheritance[$rule[2]] = $rule[0];
	}

	private function setPermission($rule) {
		if (!isset($this->permissions[$rule[1]])) {
			$this->permissions[$rule[1]] = new RootNode();
		}

		if (substr($rule[2], 0, 1) === "/") {
			$rule[2] = substr($rule[2], 1);
		}

		$this->permissions[$rule[1]]->addRule($rule[0] === "allow", $rule[2]);
	}

	function clearPermissions() {
		$this->permissionRules = "";
		$this->permissions = [];
		$this->inheritance = [];
	}
}

class MalformedRuleException extends Exception {}
